python/src/opendp/extras/mbi/_utilities/__init__.py

```diff
@@ -0,0 +1,360 @@
+            raise ValueError(f"weight ({self.weight}) must be non-negative")
+
+
+def mirror_descent(
+    domain,  # mbi.Domain
+    loss_fn,  # marginal_loss.MarginalLossFn | list[LinearMeasurement]
+    *,
+    potentials=None,  # CliqueVector | None
+):  # MarkovRandomField
+    """Fit a MarkovRandomField over the domain and loss function using mirror descent.
+    
+    Replicate the API of this function to `use other optimizers from Private-PGM <https://private-pgm.readthedocs.io/en/latest/_autosummary_output/mbi.estimation.html#module-mbi.estimation>`_.
+    """
+    from mbi.estimation import mirror_descent  # type: ignore[import-untyped,import-not-found]
+
+    return mirror_descent(domain, loss_fn, potentials=potentials)
+
+
+@dataclass(kw_only=True, frozen=True)
+class Algorithm(ABC):
+    """Base class for configuration of contingency table algorithms."""
+
+    estimator: Callable = mirror_descent
+    """Optimizer to use to fit a MarkovRandomField.
+
+    Defaults to :py:func:`opendp.extras.mbi.mirror_descent`.
+    Any function matching the signature of ``mirror_descent`` 
+    can be used to customize how the MarkovRandomField is optimized/estimated.
+    See `mbi.estimation <https://private-pgm.readthedocs.io/en/latest/_autosummary_output/mbi.estimation.html>`_ for other optimizers.
+    """
+    oneway: Literal["all", "unkeyed"] = "all"
+    """Fit one-way marginals for all columns, or only unkeyed columns."""
+    oneway_split: Optional[float] = None
+    """Proportion of budget to use for oneway release.
+    
+    When ``oneway_split`` is not set, defaults to half of the budget.
+
+    If oneway is ``unkeyed``, budget is further reduced by the proportion of columns with missing keys or cuts.
+    That is, when all columns have keys, then ``oneway_split`` is zero,
+    and when no columns have keys, then ``oneway_split`` is one-half.
+    """
+
+    def __post_init__(self):
+        if self.oneway not in {"all", "unkeyed"}:
+            raise ValueError(f'oneway ({self.oneway}) must be "all" or "unkeyed"')
+
+        if self.oneway_split is not None and not (0 <= self.oneway_split < 1):
+            raise ValueError(f"oneway_split ({self.oneway_split}) must be in [0, 1)")
+
+    @property
+    @abstractmethod
+    def make(self) -> Callable:
+        """The constructor function for fitting this algorithm."""
+        ...
+
+
+def typed_dict_domain(domains: dict[Any, Domain]) -> ExtrinsicDomain:
+    """Domain containing a fixed-key dictionary of elements, each with its own domain."""
+
+    def _member(x):
+        if not isinstance(x, dict):
+            raise ValueError("data must be a dict")
+        if set(domains.keys()) != set(x.keys()):
+            raise ValueError("data must share key-set with domain")
+        return all(domains[k].member(x[k]) for k in domains)
+
+    ident = ", ".join(f"{k}: {str(d)}" for k, d in domains.items())
+    return _extrinsic_domain(
+        identifier=f"TypedDictDomain({ident})",
```

Comment:
I'm hoping it wouldn't cause problems if the items contained commas? Nothing needs to parse this string?

Reply:
no, it's used for equality comparison

Reply:
Nothing needs to parse it, it just needs to remain unique as its used in domain equality checks.